Extern crate

bobhy opened the pull request:
Allows doc tests to freely reference external crates in `use` statements.
Avoids depending on unstable Cargo interfaces to determine dependencies by the 
simple expedient of running a `cargo build --verbose` and  observing how Cargo invoked the compiler.
Includes necessary doc changes and some gratuitous rewrites as well.

Fixes #706 and #394.
Edit: does not address issue 768.
Edit: With the latest batch of commits, all doctests in the `mdbook` project now work.

bobhy committed:
extern_args module to get proper compiler args from Cargo.

bobhy committed:
Add config.rust.package-dir and fix cargo output parser (maybe not for the last time)

bobhy committed:
Force --extern <name>=<path>.rmeta to .rlib.  Cargo sometimes tries to overoptimize compile time.

bobhy committed:
Modify user guide so code samples can reference external crates.

bobhy committed:
Update guide to describe configuring book for external crates;
Also refactor the `mdBook test` section and add a new "writing doc tests" chapter.

bobhy committed:
Fix CI nits.

bobhy committed:
Replace unstable fs::set_modified() with ad-hoc "touch" function.

bobhy committed:
Fuzzy up touch file test in case sleep() is not totally precise.

bobhy committed:
trigger rebuild if project has main.rs (but broken for other binary targets)

bobhy committed:
Make guide into workspace child so it can share same dependencies with bin.

bobhy committed:
When running build, pull dependencies from the build of the doctest crate

bobhy committed:
Compile, but do not run the preprocessor examples.

bobhy committed:
fix broken unit tests

bobhy committed:
book.tom.l must configure path to Cargo.toml, not just root folder of project: changed config rust.package-dir to rust.manifest

bobhy committed:
Doc changes to cover how to use external crates in doctests (and relation to rustdoc)

bobhy committed:
fix fmt not caught locally

bobhy committed:
Document deprecating `mdbook test -L`
and preferring [rust.manifest] over [rust.edition].

bobhy committed:
fix `mdbook test <path/to/root>;
make chap Alternate Backends pass doctest.

bobhy committed:
revert pulldown-cmark from 0.12.2 to 0.10.0 (sigh)

bobhy committed:
finalize doc updates for new feature

bobhy committed:
All doctests in mdbook project now pass.

bobhy commented:
Functionally, I think this PR is now stable and ready for review (and  general use).
The new functionality doesn't have a lot of unit tests, more could be provided per review feedback.

bobhy committed:
Provide error context if can't find or open Cargo.toml;
Prefer edition setting from Cargo.toml over book.toml; sadly, Cargo also overrides command line.

juli-99 commented on guide/src/guide/README.md:
```diff
@@ -4,4 +4,4 @@ This user guide provides an introduction to basic concepts of using mdBook.
 
 - [Installation](installation.md)
 - [Reading Books](reading.md)
-- [Creating a Book](creating.md)
```
Was this am accident?

bobhy committed:
Fix unintended change, per review

bobhy committed:
fix clippy nags in files I changed

bobhy committed:
fix unit test failure

bobhy commented:
In fixing above issues, I noticed that there were clippy nags in several files that hadn't been raised by CI.  I've fixed the ones in files I modified and could fix the remaining ones if needed, but I don't know why CI didn't flag them before this.

bobhy commented:
Will refresh the PR in a couple of weeks!

On Mon, Oct 6, 2025, 8:26 PM Berkus Decker ***@***.***> wrote:

> *berkus* left a comment (rust-lang/mdBook#2503)
> <https://github.com/rust-lang/mdBook/pull/2503#issuecomment-3373254650>
>
> @bobhy <https://github.com/bobhy> ^
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/mdBook/pull/2503#issuecomment-3373254650>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPBIZCRYZW4BLZKUE64WN33WKX45AVCNFSM6AAAAABTKIKTLOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGNZTGI2TINRVGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bobhy committed:
Try to merge (it's been a while)

bobhy committed:
Migrate changes for ExternArgs from root to crates mdbook-core and mdbook-driver.

bobhy committed:
merge upstream

bobhy committed:
recover dependency cargo_manifest

bobhy committed:
Changelog description

bobhy commented:
I'd appreciate some feedback on this PR.  It adds a kind of nice feature, but maybe needs some more work?  It's getting old, simply resolving merge conflicts...

bobhy committed:
Merge branch 'master' into extern_crate

bobhy committed:
Merge branch 'master' into extern_crate

bobhy committed:
Merge branch 'extern_crate' of https://github.com/bobhy/mdBook into extern_crate

bobhy commented:
All merged and ready for review.
Bueller?  Anyone?

GuillaumeGomez commented on crates/mdbook-core/src/utils/extern_args.rs:
```diff
@@ -0,0 +1,345 @@
+        for l in buf.lines() {
+            if let Some(_i) = l.find(" Running ") {
+                if let Some(_cn_pos) = l.find(&my_cn_arg) {
+                    let args_seg: &str = l.split('`').skip(1).take(1).collect::<Vec<_>>()[0]; // sadly, cargo decorates string with backticks
```
```suggestion
                    // sadly, cargo decorates string with backticks
                    let Some(args_seg) = l.split('`').nth(2) else {
                        builds_ignored += 1;
                        continue;
                    };
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/mdBook/2503/ec2b103e78bab036d54d6508ef8e83651d7c875d..1747565879a5b6db30e225600030b05797509274)*

bobhy replied:
It's been so long since I submitted this PR, I've forgotten a lot.  What's the point of this change?

ehuss commented:
I took a quick look at this. It looks like it is trying to parse the output from cargo which is generally intended to be human-readable only. Is there some other approach that we can take to integrate cargo with rustdoc here? For example, mdbook could be a rustdoc wrapper, or the cargo project could somehow embed the tests?

GuillaumeGomez commented:
It's possible to make `cargo` emit JSON instead and to parse this output with existing crates.

pczarn commented:
@bobhy @GuillaumeGomez I volunteer to try make this work.

GuillaumeGomez commented:
I mean... @bobhy already opened the PR so please let them finish if they want to, otherwise they will just say so.

pczarn commented on crates/mdbook-core/src/utils/extern_args.rs:
```diff
@@ -0,0 +1,345 @@
+// which is currently the MSRV for mdBook.  It is available in rust 1.76 onward.
+
+fn touch(victim: &Path) -> Result<()> {
+    let curr_content = fs::read(victim).with_context(|| "reading existing file")?;
```
```suggestion
    use std::time::SystemTime;
    let file = File::open(victim)?;
    file.set_modified(SystemTime::now())?;
    file.sync_all()?;
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/mdBook/2503/ec2b103e78bab036d54d6508ef8e83651d7c875d..1747565879a5b6db30e225600030b05797509274)*

pczarn replied:
This is merely a reminder that the function you needed became stable in std.

bobhy commented:
@pczarn If you'd like to get this cleaned up and merged, I'm happy let you run with it, and take the victory lap.  I'm available to answer questions or provide reviews on your changes.
Sounds like 2 areas of feedback:
1. replace ad-hoc "touch" function with something from `std`. 
2. Find some stable interface to parse responses from invoking Cargo (or maybe by now Cargo has a documented API for this? It has been more than a minute...)

pczarn commented:
@bobhy Happy to help. Will remember to credit your work properly.

Bet the code could always use a fresh set of eyes. It's been a while since you even touched the code, anyway.

I am not sure invoking Cargo and grabbing its output is that good of an idea. Seems cumbersome. I will see if there are any potential workarounds.

pczarn commented:
@bobhy Fun story: I am building the world's most powerful parser, and need to write lots of documentation, hence the need for this feature. Would definitely feel uneasy about pieces of code that won't be properly verified.

My section for my docs examples will be
```toml
[dependencies]
bit-vec = "0.10"
cfg = "0.10"
gearley = "0.1"
panini = "0.1"
```
